### PR TITLE
Document Telescope's Architecture

### PR DESCRIPTION
**Pull request by humphd**
In preparation for our meeting tomorrow morning, I've taken some time today to document how Telescope works.  I'd appreciate if others who have worked on it in the past could look at this and let me know what I've forgotten.

We can add to this later, but I wanted to try and talk a little about everything.

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+### Legacy Monolithic Back-end (1.0)
+
+Telescope's back-end began as a single, monolithic node.js app. During the 2.0 release, much of the back-end was split into separate microservices (see below). However, parts of the legacy back-end code are still in use, see `src/backend/*`.
+
+The current system uses the legacy backend in order to run the feed parser and feed queue, see `src/backend/feed/*`. The processed feeds and posts are then stored in Redis (cache) and Elasticsearch (indexing) databases, and various microservices use these in order to get their data.
+
+Telescope's data model is built on Feeds and Posts. A feed represents an RSS/Atom feed, and includes metadata about a particular blog (e.g., URL, author, etc) as well as URLs to individual Posts. A Post includes metadata about a particular blog post (e.g., URL, date created, date updated, etc).
+
+The legacy back-end is started using `npm start` in the root of the Telescope monorepo, and it (currently) must be run alongside the microservices. When it runs, the logs show information about feeds being parsed in real-time, which continues forever.
```
Comment (manekenpix) [inline docs/architecture.md:32]: Shouldn't this be `pnpm`?

Comment (humphd) [reply docs/architecture.md:32]: Yup, good catch.  Fixing...

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+- Auth Service (`src/api/auth`) - authentication and authorization
+- Feed-Discovery Service (`src/api/feed-discovery`) - blog feed discovery and validation
+- Image Service (`src/api/image`) - dynamic image generation for the front-end hero banner
+- Planet Service (`src/api/planet`) - recreation of the original CDOT Planet using Telescope's back-end
+- Posts Service (`src/api/posts`) - API for accessing Post and Feed data in Redis (probably not well named at this point)
+- Search Service (`src/api/search`) - API for doing searches against Elasticsearch
+- Status Service (`src/api/status`) - API for accessing Telescope status information, as well as providing the Dashboards
+- Users Service (`src/api/users`) - API for interacting with the Users data (currently in Firebase)
+- Parser Service (`src/api/parser) - feed and post parsing. Currently disabled, see <https://github.com/Seneca-CDOT/telescope/issues/2111>
```
Comment (manekenpix) [inline docs/architecture.md:52]:
```suggestion
- Parser Service (`src/api/parser`) - feed and post parsing. Currently disabled, see <https://github.com/Seneca-CDOT/telescope/issues/2111>
```

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+The microservices can be controlled either using the `docker-compose` command directly, or with the following `npm` scripts:
+
+- `npm run services:start` starts all services. You can also list one or more services to start a subset, `npm run services:start image planet`
+- `npm run services:stop` stops all services, or those specified
+- `npm run logs image` shows logs for the image service (specify whichever service you want to view)
```
Comment (manekenpix) [inline docs/architecture.md]: `pnpm` instead of `npm`?

Comment (humphd) [reply docs/architecture.md]: Yup, fixed all occurrences after you spotted the first.  New commits have updates.

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+On Windows and macOS, Docker Desktop must be running before you can call these scripts, see the [Docker documentation](docker.md).
+
+### Front-end Web App
+
+The front-end web app is written in TypeScript and uses next.js and Material UI for routing, component system, etc. The code lives in `src/web`.
+
+The web app accesses many of the microservices at runtime, and we depend on a build step to export and include the necessary runtime variables. In the front-end app, they are available at runtime via `src/web/src/config.ts`. These are populated via the `src/web/next.config.js` during build time. See [Environment Setup](environment-setup.md) for details on how to populate these with different values for local, CI, staging, or production environments. See also [Front End](front-end.md) for more details about the front-end.
+
+In front-end is run via its own container, and hosted statically by nginx. The container is defined in `src/web/Dockerfile`. Locally, it is available at <http://localhost:8000>, on staging at <https://dev.telescope.cdot.systems/>, and on production at <https://telescope.cdot.systems/>.
```
Comment (manekenpix) [inline docs/architecture.md:74]:
```suggestion
The front-end is run via its own container, and hosted statically by nginx. The container is defined in `src/web/Dockerfile`. Locally, it is available at <http://localhost:8000>, on staging at <https://dev.telescope.cdot.systems/>, and on production at <https://telescope.cdot.systems/>.
```

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+### SAML2-based authentication and authorization
+
+Telescope uses SAML2-based authentication via Seneca's Single Sign-On in combination with token-based authorization via the Auth and Users service. A good discussion of the main ideas and implementation is available in [this blog post](https://blog.humphd.org/not-so-simple-saml/) and [this Pull Request](https://github.com/Seneca-CDOT/telescope/pull/1796#issue-812477759).
+
+In development, we use a test PHP-based SAML Identity Provider to test against. See [Login](login.md).
+
+### Autodeployment Server
+
+Our release infrastructure is built primarily on GitHub Actions, see `.github/workflows/release.yml`. However, we also host our own build service at `tools/autodeployment`. When a [Release is created](release.md), the autodeployment server receives a webhook event from GitHub. We use this to trigger a new build and deploy it when complete. the build script that gets run is at `tools/autodeployment/deploy.sh`, which does a green/blue style of deployment (i.e., keep one version running while you build a new one, swap them when ready).
```
Comment (manekenpix) [inline docs/architecture.md:84]:
```suggestion
Our release infrastructure is built primarily on GitHub Actions, see `.github/workflows/release.yml`. However, we also host our own build service at `tools/autodeployment`. When a [Release is created](release.md), the autodeployment server receives a webhook event from GitHub. We use this to trigger a new build and deploy it when complete. The build script that gets run is at `tools/autodeployment/deploy.sh`, which does a green/blue style of deployment (i.e., keep one version running while you build a new one, swap them when ready).
```

**Review by manekenpix — DISMISSED** [review]

**Review by cindyledev — APPROVED** [review]

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+![Original Telescope architecture](images/initialProjectDiagram.jpeg)
+
+It shows RSS/Atom blog feeds (on the left) being downloaded (1) and then queued (2) in order to be downloaded and parsed into a database. Parsing each feed also included the idea of analysis, to look at word count, reading time, etc. This database of parsed feeds is then exposed to various clients via a REST API. We imagined being able to provide multiple clients, written for the web or mobile, but also connecting to Slack and other systems.
+
+## Current Design
+
+The following gives an overview of the current (i.e. 2.4.0) design of Telescope. There are a number of major systems that make up the app:
+
+- legacy monolothic back-end (1.0)
```
Comment (DukeManh) [inline docs/architecture.md:17]:
```suggestion
- legacy monolithic back-end (1.0)
```

**Review by DukeManh — CHANGES_REQUESTED** [review]

**Review by DukeManh — APPROVED** [review]
It's so detailed. I wonder if this can help close #2655.

**Comment by humphd** [comment]
I think #2655 could use this to help with some visualizations, which i can't do.  cc @sirinoks

**Review by manekenpix — APPROVED** [review]

**Comment by humphd** [comment]
Thanks for the reviews, everyone.

**Comment by cindyledev** [comment]
Can we reopen this lovely PR? Looks like #2655 was morphed into something unrelated to architecture

**Comment by humphd** [comment]
This got merged (manually), see https://github.com/Seneca-CDOT/telescope/blob/master/docs/architecture.md.  I don't think there's more to do?  We should file follow-ups.

**Comment by sirinoks** [comment]
I'm in process of making a clearer image to show architecture btw

**Pull request closed without merging**